### PR TITLE
Add segment selector for merged segments

### PR DESCRIPTION
**Pull request by snleee**
Once the merge lineage information is written to the property store,
broker will read the lineage and pick only segments that are required
for answering the query. This PR adds the segment selector for merged
segments.
1. Add segment selector interface
2. Add merged segment selector implementation
3. Add unit tests

### CODE REVIEW

**Comment by snleee** [comment]
#2715

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java [hunk 1]
```diff
@@ -123,16 +124,16 @@ public void computeOnExternalViewChange(String tableName, ExternalView externalV
...
       // Copy the pointer for snapshot since the pointer for segment to servers map can change at anytime
       Map<String, List<String>> segmentToServersMap = _segmentToServersMap;
 
-      // Get all existing segments
+      // Selecting segments only required for processing a query
       Set<String> segmentsToQuery = segmentToServersMap.keySet();
-
-      // TODO: add the selection logic here
+      if (segmentSelector != null) {
+        segmentsToQuery = segmentSelector.selectSegments(request, segmentsToQuery);
```
Comment (mayankshriv) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java]: Does it make sense to merge the concepts of segment selector and pruner?

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/DefaultRealtimeRoutingTableBuilder.java [hunk 2]
```diff
@@ -68,7 +69,7 @@ public void computeOnExternalViewChange(String tableName, ExternalView externalV
   }
 
   @Override
-  public Map<String, List<String>> getRoutingTable(RoutingTableLookupRequest request) {
+  public Map<String, List<String>> getRoutingTable(RoutingTableLookupRequest request, SegmentSelector segmentSelector) {
```
Comment (mayankshriv) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/DefaultRealtimeRoutingTableBuilder.java]: Should selector be passed into this api, or should it belong inside of the builder(s)?

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/selector/SegmentSelectorFactory.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+import com.linkedin.pinot.common.config.TableConfig;
+import org.apache.helix.ZNRecord;
+import org.apache.helix.store.zk.ZkHelixPropertyStore;
+
+
+/**
+ * Factory for segment selector
+ */
+public class SegmentSelectorFactory {
```
Comment (mayankshriv) [inline pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/selector/SegmentSelectorFactory.java]: This does not adhere to factory pattern, may be better to name it as SegmentSelectorProvider?

File: pinot-broker/src/test/java/com/linkedin/pinot/broker/routing/selector/MergedSegmentSelectorTest.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  @Test
+  public void testMergedSegmentSelector() throws Exception {
+    MergedSegmentSelector segmentSelector = new MergedSegmentSelector();
+
+    TableConfig tableConfig =
+        new TableConfig.Builder(CommonConstants.Helix.TableType.OFFLINE).setTableName(TEST_TABLE_NAME).build();
+    ZkHelixPropertyStore<ZNRecord> fakePropertyStore = new FakePropertyStore();
+    segmentSelector.init(tableConfig, fakePropertyStore);
+
+    // Add G0 (segment0), G1 (segment1), G2 (segment2) that represent 3 base segments
```
Comment (mayankshriv) [inline pinot-broker/src/test/java/com/linkedin/pinot/broker/routing/selector/MergedSegmentSelectorTest.java]: Could we programatically generate these trees, to improve coverage?

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java (hunk 1 above)
Comment (snleee) [reply pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/BaseRoutingTableBuilder.java]: for now, I would want to keep them separate. Once the merge feature is done, I can work on cleaning up the pruner (currently, it's closely tied with `PartitionAwareRoutingTableBuilder` and we need to decouple them)

File: pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/DefaultRealtimeRoutingTableBuilder.java (hunk 2 above)
Comment (snleee) [reply pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/DefaultRealtimeRoutingTableBuilder.java]: As discussed offline, we concluded that broker routing builder requires the re-structuring on class hierarchy. I will add the comments about it and will address later with the separate clean-up pr.

File: pinot-broker/src/test/java/com/linkedin/pinot/broker/routing/selector/MergedSegmentSelectorTest.java (hunk 4 above)
Comment (snleee) [reply pinot-broker/src/test/java/com/linkedin/pinot/broker/routing/selector/MergedSegmentSelectorTest.java]: If we generate the tree programatically, i still won't have the result to test against. I can probably come up with another implementation of the same logic directly from hashmap representation (not using tree). I will add this as a separate pr for improving testability of this segment selection logic. For now, I expanded the testing case to cover more edge cases with a fixed example.

**Pull request merged**